### PR TITLE
Turn deprecation into error for issue 313

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

**Review by wilzbach — APPROVED** [review]

**Comment by wilzbach** [comment]
Looks like more tests need updating:

```
Test runnable/testdstress.d failed. The logged output: /home/braddr/sandbox/at-client/pull-3231085-Linux_64_32/dmd/generated/linux/release/64/dmd -conf= -m32 -Irunnable -odgenerated/runnable -ofgenerated/runnable/testdstress_0 runnable/testdstress.d runnable/testdstress.d(674): Deprecation: Built-in hex string literals are deprecated, use `std.conv.hexString` instead. runnable/testdstress.d(851): Deprecation: class allocators have been deprecated, consider moving the allocation strategy outside of the class runnable/testdstress.d(173): Error: package `dstress.run` is not accessible here runnable/testdstress.d(174): Error: package `dstress.run` is not accessible here runnable/testdstress.d(175): Error: package `dstress.run` is not accessible here runnable/testdstress.d(659): Deprecation: The `delete` keyword has been deprecated. Use `object.destroy()` (and `core.memory.GC.free()` if applicable) instead. ============================== Test runnable/testdstress.d failed: 
```

File: test/compilable/testDIP37.d
Comment (JinShil) [inline test/compilable/testDIP37.d]: Rather than deleting this, should we instead add `static import pkgDIP32.datetime.common`, perhaps in a different scope?

File: test/fail_compilation/fail313.d
Comment (Geod24) [inline test/fail_compilation/fail313.d]: Remove `-de`

**Comment by Geod24** [comment]
Do you plan to do the rest of the protection code as well ? Because the time as come, and we might as well just remove the protection code instead of issuing a specific error... The former is more user friendly, though.

**Comment by RazvanN7** [comment]
@Geod24 Do you refer to : https://github.com/dlang/dmd/pull/7241 ? I was thinking of adopting that PR

File: test/compilable/testDIP37.d
Comment (RazvanN7) [reply test/compilable/testDIP37.d]: I don't see the point in that: the code will error if that different scope does not include `test7`'s scope.

**Comment by Geod24** [comment]
Forgot there was a PR for it already... Yeah that was what I was thinking about.

**Comment by RazvanN7** [comment]
Looks like FreeBSD is spuriously failing again.

File: changelog/fqn_imports_bypass_private_imports_error.dd
Comment (wilzbach) [inline changelog/fqn_imports_bypass_private_imports_error.dd]: FWIW run.dlang.io does support [HAR](https://code.dlang.org/packages/har)

File: test/compilable/testDIP37.d
Comment (wilzbach) [reply test/compilable/testDIP37.d]: The alternative would be `static assert(!__traits(compiles, ...)`

File: test/runnable/testdstress.d
Comment (wilzbach) [inline test/runnable/testdstress.d]: Why was this change necessary?

Comment (RazvanN7) [reply test/runnable/testdstress.d]: I'm not entirely sure why, but for `dstress.run.module_01` you would get that `dstress.run` is not accessible from the current module; that is probably because there is no directory structure that matches the module declaration.

Comment (wilzbach) [reply test/runnable/testdstress.d]: But the folder `run` doesn't exist either, does it?

Comment (RazvanN7) [reply test/runnable/testdstress.d]: actually its runnable

File: changelog/fqn_imports_bypass_private_imports_error.dd
Comment (wilzbach) [inline changelog/fqn_imports_bypass_private_imports_error.dd]: image

Maybe use two separate file sections in the changelog entry?

**Comment by wilzbach** [comment]
@RazvanN7 looks like this needs more work. I can reproduce failure in ae locally.

```
ae 0.0.2230: building configuration "ae-test-library"...
utils/array.d(434,18): Error: template ae.utils.array.indexOf cannot deduce function from argument types !()(string, string), candidates are:
utils/array.d(168,12):        ae.utils.array.indexOf(T, D)(in T[] arr, in D val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))) && is(typeof(arr[0] == val)))
utils/array.d(176,12):        ae.utils.array.indexOf(T)(in T[] arr, in T[] val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))))
utils/array.d(465,20): Error: template instance `ae.utils.array.skipUntil!(immutable(char), string)` error instantiating
utils/array.d(474,19):        instantiated from here: eatUntil!(immutable(char), string)
utils/array.d(434,18): Error: template ae.utils.array.indexOf cannot deduce function from argument types !()(string, char), candidates are:
utils/array.d(168,12):        ae.utils.array.indexOf(T, D)(in T[] arr, in D val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))) && is(typeof(arr[0] == val)))
utils/array.d(176,12):        ae.utils.array.indexOf(T)(in T[] arr, in T[] val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))))
utils/array.d(465,20): Error: template instance `ae.utils.array.skipUntil!(immutable(char), char)` error instantiating
utils/array.d(476,19):        instantiated from here: eatUntil!(immutable(char), char)
utils/array.d(479,38): Error: template instance `ae.utils.array.eatUntil!cast(OnEof)0.eatUntil!(immutable(char), string)` error instantiating
utils/array.d(480,43): Error: template instance `ae.utils.array.eatUntil!cast(OnEof)1.eatUntil!(immutable(char), string)` error instantiating
utils/text/package.d(71,18): Error: template instance `ae.utils.text.eatLine!(immutable(char))` error instantiating
/home/seb/dlang/dmd/generated/linux/release/64/dmd failed with exit code 1.
```

**Comment by wilzbach** [comment]
Looked a bit closer and the `(string, string)` part is the reason for the failure.
The two overloads have `!isSomeChar!T`, s.t. they don't conflict with `std.string`. However, the import of `std.string` is missing and the errors are about calls to the `string` variant.
This would solve the problem:

```d
public import std.string;
```

But as there was never a deprecation triggered, I'm not sure whether we can go ahead with this directly.
CC @CyberShadow what do you think? It's your library that is blocking this PR ;-)

Maybe related: https://github.com/CyberShadow/ae/pull/33

**Comment by CyberShadow** [comment]
> But as there was never a deprecation triggered, I'm not sure whether we can go ahead with this directly.

Yes, that bug ought to be fixed first.

**Comment by JinShil** [comment]
> Yes, that bug ought to be fixed first.

I assume you're referring to a DMD bug.  What specifically is the bug?  Is it in bugzilla yet?  Do you have a minimal test case to trigger the bug?

**Comment by CyberShadow** [comment]
> I assume you're referring to a DMD bug. What specifically is the bug?

There seems to be a case for which a deprecation should have been emitted, but wasn't. This pull request also turns that case into an error, thus breaking code with no prior warning. See https://github.com/CyberShadow/ae/pull/33 .

> Is it in bugzilla yet? Do you have a minimal test case to trigger the bug?

Not that I know of.

**Comment by JinShil** [comment]
> Is it in bugzilla yet? Do you have a minimal test case to trigger the bug?

https://issues.dlang.org/show_bug.cgi?id=19107

File: test/runnable/testdstress.d
Comment (JinShil) [reply test/runnable/testdstress.d]: That module declaration looks like a leftover from when code was copied into the `test/runnable` directory.  Can't we just get rid of it?

**Comment by JinShil** [comment]
Is Jenkins dead now?

**Comment by wilzbach** [comment]
Yes.

**Comment by JinShil** [comment]
> Ocean (genuine DMD segfault - I wonder why the stacktrace doesn't have line numbers though.)

Might be fixed with https://github.com/dlang/druntime/pull/2278

**Comment by ibuclaw** [comment]
> Setting this to blocked, s.t. we remember that this can't be merged before the 2.082 has happened (~15.8).

I do believe that we can unblock this now.

**Comment by ZombineDev** [comment]
@RazvanN7 rebase pls

**Comment by RazvanN7** [comment]
@ZombineDev  Done.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by wilzbach** [comment]
The ae failure at buildkite still looks genuine.

**Comment by ZombineDev** [comment]
@CyberShadow can you take a look at the ae errors?

**Comment by CyberShadow** [comment]
I literally can't. Could you please repaste the log?

If I build this PR with Digger, ae builds and passes tests successfully with the resulting DMD.

**Comment by Geod24** [comment]
```

~> bash
--
  | Generating test runner configuration 'ae-test-library' for 'library' (sourceLibrary).
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Excluding package.d file from test due to https://issues.dlang.org/show_bug.cgi?id=11847
  | Performing "unittest" build using dmd for x86_64.
  | ae 0.0.2281: building configuration "ae-test-library"...
  | utils/array.d(434,18): Error: template `ae.utils.array.indexOf` cannot deduce function from argument types `!()(string, string)`, candidates are:
  | utils/array.d(168,12):        `ae.utils.array.indexOf(T, D)(in T[] arr, in D val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))) && is(typeof(arr[0] == val)))`
  | utils/array.d(176,12):        `ae.utils.array.indexOf(T)(in T[] arr, in T[] val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))))`
  | utils/array.d(465,20): Error: template instance `ae.utils.array.skipUntil!(immutable(char), string)` error instantiating
  | utils/array.d(474,19):        instantiated from here: `eatUntil!(immutable(char), string)`
  | utils/array.d(434,18): Error: template `ae.utils.array.indexOf` cannot deduce function from argument types `!()(string, char)`, candidates are:
  | utils/array.d(168,12):        `ae.utils.array.indexOf(T, D)(in T[] arr, in D val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))) && is(typeof(arr[0] == val)))`
  | utils/array.d(176,12):        `ae.utils.array.indexOf(T)(in T[] arr, in T[] val) if (!isSomeChar!T && is(typeof(arr.countUntil(val))))`
  | utils/array.d(465,20): Error: template instance `ae.utils.array.skipUntil!(immutable(char), char)` error instantiating
  | utils/array.d(476,19):        instantiated from here: `eatUntil!(immutable(char), char)`
  | utils/array.d(479,38): Error: template instance `ae.utils.array.eatUntil!cast(OnEof)0.eatUntil!(immutable(char), string)` error instantiating
  | utils/array.d(480,43): Error: template instance `ae.utils.array.eatUntil!cast(OnEof)1.eatUntil!(immutable(char), string)` error instantiating
  | utils/graphics/color.d(498,45): Deprecation: integral promotion not done for `~b`, use '-transition=intpromote' switch or `~cast(int)(b)`
  | utils/graphics/color.d(498,45): Deprecation: integral promotion not done for `~b`, use '-transition=intpromote' switch or `~cast(int)(b)`
  | utils/graphics/color.d(498,45): Deprecation: integral promotion not done for `~b`, use '-transition=intpromote' switch or `~cast(int)(b)`
  | utils/graphics/color.d(498,45): Deprecation: integral promotion not done for `~b`, use '-transition=intpromote' switch or `~cast(int)(b)`
  | utils/graphics/color.d(498,45): Deprecation: integral promotion not done for `~b`, use '-transition=intpromote' switch or `~cast(int)(b)`
  | utils/text/package.d(71,18): Error: template instance `ae.utils.text.eatLine!(immutable(char))` error instantiating
  | dmd failed with exit code 1.
  | 🚨 Error: The command exited with status 2


```

**Comment by CyberShadow** [comment]
Those errors don't make sense, look at the surrounding code.

> If I build this PR with Digger, ae builds and passes tests successfully with the resulting DMD.

Oops, forgot to add `--compiler=dmd`.

**Comment by CyberShadow** [comment]
Pushed fix & tag.

**Review by jacob-carlborg — APPROVED** [review]

**Review by PetarKirov — APPROVED** [review]

**Pull request merged**
